akka-actor/src/main/scala-2.13+/akka/util/ByteString.scala

```diff
@@ -361,10 +367,22 @@ object ByteString {
     def asByteBuffers: scala.collection.immutable.Iterable[ByteBuffer] = List(asByteBuffer)
 
     override def decodeString(charset: String): String =
-      new String(if (length == bytes.length) bytes else toArray, charset)
+      if (isEmpty) ""
+      else new String(bytes, startIndex, length, charset)
 
     override def decodeString(charset: Charset): String = // avoids Charset.forName lookup in String internals
-      new String(if (length == bytes.length) bytes else toArray, charset)
+      if (isEmpty) ""
+      else new String(bytes, startIndex, length, charset)
```

Comment:
This implementation avoids having to allocate a new array in `toArray` which would then immediately be thrown away by the String constructor, which wants its own array.

Also, arguably the implementation is clearer, since it doesn't use the implication that startIndex == 0 when length == bytes.length